### PR TITLE
[SPARK-23631][ML][PySpark] Add summary to RandomForestClassificationModel

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Add summary to RandomForestClassificationModel...

### Why are the changes needed?
so user can get a summary of this classification model, and retrieve common metrics such as accuracy, weightedTruePositiveRate, roc (for binary), pr curves (for binary), etc.

### Does this PR introduce _any_ user-facing change?
Yes
```
RandomForestClassificationModel.summary
RandomForestClassificationModel.evaluate
```

### How was this patch tested?
Add new tests

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/classification/RandomForestClassifier.scala [hunk 1]
```diff
@@ -166,7 +166,35 @@ class RandomForestClassifier @Since("1.4.0") (
...
+
+    val (summaryModel, probabilityColName, predictionColName) = model.findSummaryModel()
+    val rfSummary = if (numClasses <= 2) {
+      new BinaryRandomForestClassificationTrainingSummaryImpl(
+        summaryModel.transform(dataset),
+        probabilityColName,
+        predictionColName,
+        $(labelCol),
+        weightColName,
+        Array(0.0))
```
Comment (huaxingao) [inline mllib/src/main/scala/org/apache/spark/ml/classification/RandomForestClassifier.scala:188]: for non iterative algorithm, set objectiveHistory to Array(0.0).

**Comment by zhengruifeng** [comment]
It seems that you need to add summary in ml one by one.
Is it possible to add it as a generic function in `Classifier`/`Regressor` or even in `Predictor`?

**Review by srowen — APPROVED** [review]
Looks good to me.

**Comment by huaxingao** [comment]
@zhengruifeng 
Thanks for your comments.
Seems I have to add summary to each of the Classifiers one by one. Using this PR as an example, I think I need to have the following in RandomForestclassifier.scala, to be consistent with the implementation of LogisticRegressionTrainingSummary.
```
sealed trait RandomForestClassificationSummary extends ClassificationSummary 

sealed trait RandomForestClassificationTrainingSummary extends RandomForestClassificationSummary
  with TrainingSummary

sealed trait BinaryRandomForestClassificationSummary extends BinaryClassificationSummary

sealed trait BinaryRandomForestClassificationTrainingSummary extends
  BinaryRandomForestClassificationSummary with RandomForestClassificationTrainingSummary

private class RandomForestClassificationTrainingSummaryImpl

private class RandomForestClassificationSummaryImpl

private class BinaryRandomForestClassificationTrainingSummaryImpl

private class BinaryRandomForestClassificationSummaryImpl
```

**Comment by srowen** [comment]
Merged to master

**Comment by huaxingao** [comment]
Thanks! @srowen @zhengruifeng

**Pull request closed without merging**